jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java

```diff
@@ -21,36 +21,31 @@
 import java.nio.channels.ReadableByteChannel;
 import java.nio.file.Path;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.Comparator;
 import java.util.HashSet;
 import java.util.List;
+import java.util.stream.Collectors;
 
 import org.eclipse.jetty.util.IO;
-import org.eclipse.jetty.util.StringUtil;
 import org.eclipse.jetty.util.URIUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
- * A collection of resources (dirs).
- * Allows webapps to have multiple (static) sources.
+ * A collection of Resources.
+ * Allows webapps to have multiple sources.
  * The first resource in the collection is the main resource.
  * If a resource is not found in the main resource, it looks it up in
- * the order the resources were constructed.
+ * the order the provided in the constructors
  */
-public class ResourceCollection extends Resource
+public class ResourceCollection extends Resource implements AutoCloseable
 {
-    private List<Resource> _resources;
+    private static final Logger LOG = LoggerFactory.getLogger(ResourceCollection.class);
 
-    /**
-     * Instantiates an empty resource collection.
-     * <p>
-     * This constructor is used when configuring jetty-maven-plugin.
-     */
-    public ResourceCollection()
-    {
-        _resources = new ArrayList<>();
-    }
+    private final List<Resource> _resources;
+    private final List<Mount> _mounted = new ArrayList<>();
```

Comment:
I disagree, as that puts the mounting logic everywhere.
It needs to be centralized and done properly.

Ideally, the `ContextHandler.baseResource` (and `ServletContextHandler`, and `WebAppContext`) will all be a `ResourceCollection` (like every `WebAppContext` instance is since Jetty 9.x).

We don't want each Configuration managing mounting, and the ContextHandler, and the ServletContextHandler, and the WebAppContext, and the DefaultServlet, and the ResourceService, and the ResourceHandler, and the QuickStart (and that's just to name a few).

It needs to be centralized.

There's also a different discussion on how `ContextHandler.getBaseResource()` behaves, as it's been broken for MultiRelease jar files since they were introduced as a concept in Java.

Take this example.

Call to `WebAppContext.getBaseResource().resolve("WEB-INF")` returns what?
(about 98% of the time this is a `ResourceCollection` courtesy of 3 different `Configuration` behaviors).
This could return a `WEB-INF/lib/foo.jar!/WEB-INF/` or even `WEB-INF/lib/foo.jar!/META-INF/versions/11/WEB-INF`.
For users of `getBaseResource().resolve(String)`, they need more control over what to search. (the first entry, which might be the WAR? or all entries?)
This also has impacts on TLD searches, and even QuickStart.

If we had `PathCollection` still as the base for a ContextHandler, then you can `pathCollection.stream().filter(FileID::isWar).filter(p->p.getFileName().endsWith(".tld"))...`
Which you cannot do in `Resource` or `ResourceCollection`.
Instead, you have to ask, every time you call `.getBaseResource()`, are you a collection? what's the first entry? is it a war? lets search that only.  Oops, if there was a TLD search, then the WAR might not be first, as that was created by a Set from one of the Configuration classes.
Wait, how is an entry in the `ResourceCollection` also another `ResourceCollection`? (this happens with the Configuration classes too!)